Extracting velocities/forces from TRR files (hidden API)

dwhswenson opened the pull request:
This PR is related to the questions in #1230, but doesn't attempt to change the core MDTraj API. Instead, this just provides options (within the "hidden" methods like `_read` and `_write`) to read/write velocities and forces in TRR files.

Basically, I wanted to have MDTraj's seek capabilities in TRR files, plus the Cython speed, without having to fork and maintain it myself (or add another requirement). So I've tried to implement it so that I can get the extra features I'd like, without my code getting in the way. Call them hidden features, or preparation for solving #1230 in the future.

One thing I would like to change, but will do as you prefer: Right now, the number of return values for `_read` depends on the values of the parameters `get_velocities` and `get_forces`. If both are false, it returns 5 values. If one is true, it returns 6 values. If both are true, it returns 7 values.

That's ugly, but was done with the intent of minimally modifying any other code. If you'd prefer, I'd be happy to make it always return 7 values, and with `None` in the location of anything that isn't retrieved. I consider methods beginning with `_` to be in hidden APIs, and therefore subject to change without worrying about backward compatibility. But you might feel differently, so I maintained backward compatibility.

rmcgibbo commented:
Nice work.

I think it makes a lot of sense to add this ability to the low-level TRRTrajectoryFile whether or not it ends up getting exposed in the Trajectory class.

As the modifications are to method with an underscore, prefix, I think it's totally fine to change the API to the (more natural) unconditional 7-elements return value. The tradeoff is between code simplicity and backward compatibility, and I'm more than happy to favor simplicity in this case.


> On Jun 2, 2017, at 6:54 PM, David W.H. Swenson <notifications@github.com> wrote:
> 
> This PR is related to the questions in #1230, but doesn't attempt to change the core MDTraj API. Instead, this just provides options (within the "hidden" methods like _read and _write) to read/write velocities and forces in TRR files.
> 
> Basically, I wanted to have MDTraj's seek capabilities in TRR files, plus the Cython speed, without having to fork and maintain it myself (or add another requirement). So I've tried to implement it so that I can get the extra features I'd like, without my code getting in the way. Call them hidden features, or preparation for solving #1230 in the future.
> 
> One thing I would like to change, but will do as you prefer: Right now, the number of return values for _read depends on the values of the parameters get_velocities and get_forces. If both are false, it returns 5 values. If one is true, it returns 6 values. If both are true, it returns 7 values.
> 
> That's ugly, but was done with the intent of minimally modifying any other code. If you'd prefer, I'd be happy to make it always return 7 values, and with None in the location of anything that isn't retrieved. I consider methods beginning with _ to be in hidden APIs, and therefore subject to change without worrying about backward compatibility. But you might feel differently, so I maintained backward compatibility.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/mdtraj/mdtraj/pull/1271
> 
> Commit Summary
> 
> Start adding vel/force reading in TRR
> Merge branch 'master' into gromacs_vel
> Internal read/write of velocities, forces in TRR
> Merge remote-tracking branch 'upstream/master' into gromacs_vel
> First unit test works
> Remaining unit tests
> File Changes
> 
> M mdtraj/formats/xtc/trr.pyx (58)
> M mdtraj/tests/test_trr.py (106)
> Patch Links:
> 
> https://github.com/mdtraj/mdtraj/pull/1271.patch
> https://github.com/mdtraj/mdtraj/pull/1271.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dwhswenson commented:
Thanks @rmcgibbo. I switched to `._read` always returning a 7-tuple, and also added support for when `atom_indices` is used.

I think this is ready for review. It looks like the tests that it fails are a general problem right now with Linux builds on Travis (AppVeyor and Travis/Mac tests pass).

Summary of what's in here:

* `TRRTrajectoryFile._read` always returns 7 items: `xyz`, `time`, `step`, `box`, `lambda`, `vel`, `forces`.
* `TRRTrajectoryFile._read` takes boolean arguments `get_velocities` and `get_forces`: if `True`, it returns the numpy arrays in the slots for those items; if `False`, those slots in the return tuple are `None`. Default is `False` for both.
* If `get_velocities=True` is passed to `._read`, but no velocities are stored in the TRR file, it raises a `RuntimeError` (similarly for forces).
* Support for all of this whether `atom_indices` are specified to `._read` or not.
* New static method `._check_has_velocities_forces` checks for presence of velocities/forces in a file (returning 2-tuple `(has_velocities, has_forces)`, each nonzero and positive if answer is yes.)
* `TRRTrajectoryFile._write` can optionally take `vel` and `force` (numpy arrays).
* New unit tests cover all this behavior.

Also some changes to make a few parts more readable on 80 columns, and shifted the zero-filled allocation of buffers (only used if `atom_indices` is not `None`) to only occur when `atom_indices` is not `None`.

mpharrigan commented:
rebased in #1294
